src/ingress.py

```diff
@@ -12,12 +12,19 @@
 class Observer(ops.Object):
```

Comment:
I'm wondering if the `Observer` for ingress is strictly necessary - Could we instantiate IngressPerAppRequirer in agent observer or in charm class and pass it to agent observer?
My understanding is that an observer is used to simplify the `framework.observe` in the main charm class. Since this doesn't quite observer anything here, I'm wondering if removing this observer class would simplify the code. 
What do you think? :D

Reply:
Or we could move the agent_discovery_url and the observer events here so that the code related to ingress changes are more cohesive

Reply:
Sounds reasonable, currently the agent observer needs to use the ingress Observer in its constructor to stick with the Architecture guidance and avoid cyclic dependency, so cleaning up the ingress observer would simplify this a bit. But merging the 2 observer does raise the question about separation of concerns. 

I'm marking this as resolved, let's continue the discussion and I'll open a follow-up PR if needed.